### PR TITLE
Add svtplay series playlist support

### PR DESCRIPTION
**Pull request by wader**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Add support for SVT play series as playlist

### CODE REVIEW

**Comment by dstftw** [comment]
Core tests broken.

File: youtube_dl/extractor/svt.py [hunk 1]
```diff
@@ -189,3 +190,53 @@ def _real_extract(self, url):
                     r'\s*\|\s*.+?$', '',
                     info_dict.get('episode') or self._og_search_title(webpage))
             return info_dict
+
+
+class SVTPlaylistIE(InfoExtractor):
+    IE_DESC = 'SVT Play serie'
+    _VALID_URL = r'https?://(?:www\.)?svtplay\.se/(?P<id>[^/]+)($|\?.*)'
```
Comment (dstftw) [inline youtube_dl/extractor/svt.py]: No. Override `suitable` instead.

File: youtube_dl/extractor/svt.py [hunk 2]
```diff
@@ -189,3 +190,53 @@ def _real_extract(self, url):
...
+    IE_NAME = 'svtplay:serie'
+    _TESTS = [{
+        'url': 'https://www.svtplay.se/rederiet',
+        'info_dict': {
+            'id': 'rederiet',
+            'title': 'Rederiet',
+            'description': 'Slaget om Östersjön är en kamp om miljarder, på liv och död. Under seriens gång för Dahlénrederiet en ständig kamp mot  andra rederier om farlederna mellan Sverige och Finland.  Dramaserien om familjen Dahlén, båten M/S Freja och dess besättning är den längsta tv-serien i svensk historia. Totalt sändes Rederiet i 20 säsonger mellan åren 1992-2002. I rollerna: Johannes Brost (Torbjörn "Joker" Jonasson), Hans V Engström (Uno Kronkvist), Gösta Prüzelius (Reidar Dahlén), Suzanne Reuter (Renate Dahlén), Lena Strömdahl (Yvonne Dahlén), Per Holmberg (Henrik Bjurhed), Kenneth Söderman (Tony Sjögren), Peter Harryson (Pehr Silver), Bert-Åke Varg (Gustav Sjögren), Bert-Åke Varg (Gustav Sjögren), Bengt Bauler (Carl Ericson), Ola Forssmed (Micki Sandell) m.fl.',
```
Comment (dstftw) [inline youtube_dl/extractor/svt.py]: `md5:...`.

File: youtube_dl/extractor/svt.py [hunk 3]
```diff
@@ -189,3 +190,53 @@ def _real_extract(self, url):
...
+        },
+        'playlist_mincount': 318,
+    }]
+
+    def _real_extract(self, url):
+        mobj = re.match(self._VALID_URL, url)
+        id = mobj.group('id')
```
Comment (dstftw) [inline youtube_dl/extractor/svt.py]: `_match_id`. Do not shadow built-in names.

File: youtube_dl/extractor/svt.py [hunk 4]
```diff
@@ -189,3 +190,53 @@ def _real_extract(self, url):
...
+
+        root_json = self._search_regex(
+            'root\\[\'__svtplay\'\\]\\s*=(.*);\\s*root\\[\'__env\'\\]',
+            page, "root", default=None, fatal=True)
+        root = self._parse_json(root_json, id)
+
+        metadata = root["metaData"]
+
+        entries = []
+        for season in root["relatedVideoContent"]["relatedVideosAccordion"]:
```
Comment (dstftw) [inline youtube_dl/extractor/svt.py]: Single quotes.

File: youtube_dl/extractor/svt.py [hunk 5]
```diff
@@ -189,3 +190,53 @@ def _real_extract(self, url):
...
+            for video in season["videos"]:
+                contentUrl = video.get("contentUrl")
+                if contentUrl is None:
```
Comment (dstftw) [inline youtube_dl/extractor/svt.py]: `if not content_url:`.

File: youtube_dl/extractor/svt.py [hunk 6]
```diff
@@ -189,3 +190,53 @@ def _real_extract(self, url):
...
+        id = mobj.group('id')
+
+        page = self._download_webpage(
+            url, None,
+            note='Downloading serie page',
+            errnote='unable to fetch serie page', fatal=True)
+
+        root_json = self._search_regex(
+            'root\\[\'__svtplay\'\\]\\s*=(.*);\\s*root\\[\'__env\'\\]',
+            page, "root", default=None, fatal=True)
```
Comment (dstftw) [inline youtube_dl/extractor/svt.py]: default and fatal are not used together. `True` is default.

File: youtube_dl/extractor/svt.py [hunk 7]
```diff
@@ -189,3 +190,53 @@ def _real_extract(self, url):
...
+    }]
+
+    def _real_extract(self, url):
+        mobj = re.match(self._VALID_URL, url)
+        id = mobj.group('id')
+
+        page = self._download_webpage(
+            url, None,
+            note='Downloading serie page',
+            errnote='unable to fetch serie page', fatal=True)
```
Comment (dstftw) [inline youtube_dl/extractor/svt.py]: `True` is default.

**Comment by wader** [comment]
@dstftw now core tests are green. thanks for the feedback

File: youtube_dl/extractor/svt.py [hunk 8]
```diff
@@ -189,3 +190,57 @@ def _real_extract(self, url):
...
+        'playlist_mincount': 318,
+    }]
+
+    @classmethod
+    def suitable(cls, url):
+        return False if SVTIE.suitable(url) or SVTPlayIE.suitable(url) else super(SVTPlaylistIE, cls).suitable(url)
+
+    def _real_extract(self, url):
+        mobj = re.match(self._VALID_URL, url)
+        _match_id = mobj.group('id')
```
Comment (dstftw) [inline youtube_dl/extractor/svt.py]: `self._match_id`.

File: youtube_dl/extractor/svt.py [hunk 9]
```diff
@@ -189,3 +190,57 @@ def _real_extract(self, url):
...
+
+        page = self._download_webpage(
+            url, None,
```
Comment (dstftw) [inline youtube_dl/extractor/svt.py]: `None` is not an id.

File: youtube_dl/extractor/svt.py [hunk 10]
```diff
@@ -189,3 +190,57 @@ def _real_extract(self, url):
...
+            note='Downloading serie page',
+            errnote='unable to fetch serie page')
+
+        root_json = self._search_regex(
+            'root\\[\'__svtplay\'\\]\\s*=(.*);\\s*root\\[\'__env\'\\]',
+            page, "root")
+        root = self._parse_json(root_json, _match_id)
+
+        metadata = root["metaData"]
```
Comment (dstftw) [inline youtube_dl/extractor/svt.py]: Quotes.

File: youtube_dl/extractor/svt.py [hunk 11]
```diff
@@ -189,3 +190,57 @@ def _real_extract(self, url):
...
+
+        entries = []
+        for season in root['relatedVideoContent']['relatedVideosAccordion']:
+            for video in season['videos']:
+                contentUrl = video.get('contentUrl')
+                if not contentUrl:
```
Comment (dstftw) [inline youtube_dl/extractor/svt.py]: No camelCase.

**Comment by wader** [comment]
@dstftw Sorry for all the small messups. I've fixed some more stuff also

File: youtube_dl/extractor/svt.py [hunk 12]
```diff
@@ -189,3 +190,56 @@ def _real_extract(self, url):
...
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+
+        page = self._download_webpage(
+            url, video_id,
+            note='Downloading serie page',
+            errnote='unable to fetch serie page')
+
+        root_json = self._search_regex(
+            'root\\[\'__svtplay\'\\]\\s*=(.*);\\s*root\\[\'__env\'\\]',
```
Comment (dstftw) [inline youtube_dl/extractor/svt.py]: Use raw strings.

Comment (dstftw) [inline youtube_dl/extractor/svt.py]: Capturing empty string is senseless. `\n` instead of `root[__env]`.

File: youtube_dl/extractor/svt.py [hunk 13]
```diff
@@ -189,3 +190,56 @@ def _real_extract(self, url):
...
+                        urljoin(url, content_url),
+                        ie=SVTPlayIE.ie_key(),
+                        video_title=video.get('title')
+                    ))
+
+        return self.playlist_result(
+            entries,
+            video_id,
+            metadata.get('title'),
+            metadata.get('description'))
```
Comment (dstftw) [inline youtube_dl/extractor/svt.py]: Eliminate excessive verbosity.

File: youtube_dl/extractor/svt.py [hunk 14]
```diff
@@ -189,3 +190,56 @@ def _real_extract(self, url):
...
+            url, video_id,
+            note='Downloading serie page',
+            errnote='unable to fetch serie page')
+
+        root_json = self._search_regex(
+            'root\\[\'__svtplay\'\\]\\s*=(.*);\\s*root\\[\'__env\'\\]',
+            page, 'root')
+        root = self._parse_json(root_json, video_id)
+
+        metadata = root['metaData']
```
Comment (dstftw) [inline youtube_dl/extractor/svt.py]: Breaks extraction if missing.

File: youtube_dl/extractor/svt.py [hunk 15]
```diff
@@ -189,3 +190,56 @@ def _real_extract(self, url):
                     r'\s*\|\s*.+?$', '',
                     info_dict.get('episode') or self._og_search_title(webpage))
             return info_dict
+
+
+class SVTPlaylistIE(InfoExtractor):
+    IE_DESC = 'SVT Play serie'
+    _VALID_URL = r'https?://(?:www\.)?svtplay\.se/(?P<id>[^/]+)'
```
Comment (dstftw) [inline youtube_dl/extractor/svt.py]: Captures `?#&` ending as id.

File: youtube_dl/extractor/svt.py [hunk 16]
```diff
@@ -189,3 +190,56 @@ def _real_extract(self, url):
...
+        root_json = self._search_regex(
+            'root\\[\'__svtplay\'\\]\\s*=(.*);\\s*root\\[\'__env\'\\]',
+            page, 'root')
+        root = self._parse_json(root_json, video_id)
+
+        metadata = root['metaData']
+
+        entries = []
+        for season in root['relatedVideoContent']['relatedVideosAccordion']:
+            for video in season['videos']:
```
Comment (dstftw) [inline youtube_dl/extractor/svt.py]: Breaks if no videos in season.

**Comment by wader** [comment]
Ready for merge?

File: youtube_dl/extractor/svt.py [hunk 17]
```diff
@@ -189,3 +190,67 @@ def _real_extract(self, url):
...
+            errnote='unable to fetch serie page')
+
+        root_json = self._search_regex(
+            r'root\[\'__svtplay\'\]\s*=(.+);\n',
+            page, 'root')
+        root = self._parse_json(root_json, video_id)
+
+        metadata = root.get('metaData')
+        if not metadata:
+            raise ExtractorError('root.metaData missing')
```
Comment (dstftw) [inline youtube_dl/extractor/svt.py]: Remove.

File: youtube_dl/extractor/svt.py [hunk 18]
```diff
@@ -189,3 +190,67 @@ def _real_extract(self, url):
...
+
+        related_video_content = root.get('relatedVideoContent')
+        if not related_video_content:
+            raise ExtractorError('root.relatedVideoContent missing')
```
Comment (dstftw) [inline youtube_dl/extractor/svt.py]: Remove.

File: youtube_dl/extractor/svt.py [hunk 19]
```diff
@@ -189,3 +190,67 @@ def _real_extract(self, url):
...
+
+        related_videos_accordion = related_video_content.get('relatedVideosAccordion')
+        if not related_videos_accordion:
+            raise ExtractorError('root.relatedVideosAccordion missing')
```
Comment (dstftw) [inline youtube_dl/extractor/svt.py]: Remove.

File: youtube_dl/extractor/svt.py [hunk 20]
```diff
@@ -189,3 +190,67 @@ def _real_extract(self, url):
...
+
+        entries = []
+        for season in related_videos_accordion:
+            videos = season.get('videos')
+            if not videos:
```
Comment (dstftw) [inline youtube_dl/extractor/svt.py]: Should check for a list.

File: youtube_dl/extractor/svt.py [hunk 21]
```diff
@@ -189,3 +190,67 @@ def _real_extract(self, url):
...
+                continue
+
+            for video in videos:
+                content_url = video.get('contentUrl')
+                if not content_url:
```
Comment (dstftw) [inline youtube_dl/extractor/svt.py]: Check for `compat_str`.

**Comment by wader** [comment]
@dstftw  Thanks, will fix. Is there any extractor that follows all guide and style -lines? I’ve looked at other extractors for how to handle errors etc but they might be outdated?

**Comment by dstftw** [comment]
Maybe. Try looking at some newly added extractors. The main point of all these checks is to foresee and minimize potential breakages due to unexpected data.

**Comment by wader** [comment]
@dstftw Fixed. Not sure i follow if the idea is to return empty results on unexpected data or to crash with as useful error as possible? for expected errors like auth error etc `ExtractorError` should be used? Looking at some of the new extractors they seem to poke around in JSON without much errors checks, like https://github.com/rg3/youtube-dl/blob/master/youtube_dl/extractor/tennistv.py#L94

Let me know if i should add more check

**Comment by dstftw** [comment]
The idea is not to crash on any missing optional data and crash on missing mandatory data. What do you mean by useful error? Crashing with stacktrace on `[]` operator accessing mandatory key in dict is much more useful than throwing something like `root.relatedVideosAccordion missing`. This information is for developer in the first place not end user. Both means nothing for end user but the former is much more beneficial for developer since it shows exact place of breakage.
tennistv is really bad in terms of robustness and will break on any minor change/unexpected data.

**Comment by wader** [comment]
@dstftw Yes i agree. Handle optional and crash on unexpected with pyhon stacktrace. So i should change `metadata = root.get('metaData')` to `metadata = ['metaData']` and same for `season.get('videos')` and  `video.get('contentUrl')`?  They are not optional and i'm kind of certain they should always exist unless there has been some major change to the site.

**Comment by dstftw** [comment]
No. You use `metaData` only for playlist title and description both of which are considered optional. `videos` and `contentUrl` should not be changed to mandatory as well since you process list of such data: some of which may be present, some - not. Even single one extracted `contentUrl` is better than a crash without anything.

**Comment by wader** [comment]
@dstftw Ok thanks. I've updated the PR now. Only change was `metadata = root.get('metaData', {})` to make `{}` the default value to make `metadata.get('title')` etc don't crash later on.

**Comment by wader** [comment]
@dstftw Thanks for the patience. Next PR will go smoother i think 👍

**Pull request closed without merging**